Lib/profiling/sampling/flamegraph.js

```diff
@@ -0,0 +1,442 @@
+function main() {
+  const data = {{FLAMEGRAPH_DATA}}
+
+  if (typeof flamegraph === "undefined") {
+    console.error("d3-flame-graph library not loaded");
+    document.getElementById("chart").innerHTML =
+      '<h2 style="text-align: center; color: #d32f2f;">Error: d3-flame-graph library failed to load</h2>';
+    throw new Error("d3-flame-graph library failed to load");
+  }
+
+  const pythonTooltip = flamegraph.tooltip.defaultFlamegraphTooltip();
+
+  pythonTooltip.show = function (d, element) {
+    if (!this._tooltip) {
+      this._tooltip = d3
+        .select("body")
+        .append("div")
+        .attr("class", "python-tooltip")
+        .style("position", "absolute")
+        .style("padding", "20px")
+        .style("background", "white")
+        .style("color", "#2e3338")
+        .style("border-radius", "8px")
+        .style("font-size", "14px")
+        .style("border", "1px solid #e9ecef")
+        .style("box-shadow", "0 8px 30px rgba(0, 0, 0, 0.15)")
+        .style("z-index", "1000")
+        .style("pointer-events", "none")
+        .style("font-weight", "400")
+        .style("line-height", "1.5")
+        .style("max-width", "500px")
+        .style("font-family", "'Source Sans Pro', sans-serif")
+        .style("opacity", 0);
+    }
+
+    const timeMs = (d.data.value / 1000).toFixed(2);
+    const percentage = ((d.data.value / data.value) * 100).toFixed(2);
+    const calls = d.data.calls || 0;
+    const childCount = d.children ? d.children.length : 0;
+    const source = d.data.source;
+
+    // Create source code section if available
+    let sourceSection = "";
+    if (source && Array.isArray(source) && source.length > 0) {
+      const sourceLines = source
+        .map(
+          (line) =>
+            `<div style="font-family: 'SF Mono', 'Monaco', 'Consolas', monospace; font-size: 12px; color: ${line.startsWith("→") ? "#3776ab" : "#5a6c7d"}; white-space: pre; line-height: 1.4; padding: 2px 0;">${line.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;")}</div>`,
+        )
+        .join("");
+
+      sourceSection = `
+                    <div style="margin-top: 16px; padding-top: 12px; border-top: 1px solid #e9ecef;">
+                        <div style="color: #3776ab; font-size: 13px; margin-bottom: 8px; font-weight: 600;">📄 Source Code:</div>
+                        <div style="background: #f8f9fa; border: 1px solid #e9ecef; border-radius: 6px; padding: 12px; max-height: 150px; overflow-y: auto;">
+                            ${sourceLines}
+                        </div>
+                    </div>
+                `;
+    } else if (source) {
+      // Show debug info if source exists but isn't an array
+      sourceSection = `
+                    <div style="margin-top: 16px; padding-top: 12px; border-top: 1px solid #e9ecef;">
+                        <div style="color: #d32f2f; font-size: 13px; margin-bottom: 8px; font-weight: 600;">🐛 Debug - Source data type: ${typeof source}</div>
```

Comment:
Per above, also a suggestion to emphasise 'debug' in the output
```suggestion
                        <div style="color: #d32f2f; font-size: 13px; margin-bottom: 8px; font-weight: 600;">[Debug] Source data type: ${typeof source}</div>
```